packages/react-aria-components/src/Tree.tsx

```diff
@@ -71,37 +139,179 @@ class TreeCollection<T> implements ICollection<Node<T>> {
   }
 
   at(idx: number) {
-    return this.flattenedRows[idx];
+    let keyMap = this.keyMap;
+    let expandedKeys = this.expandedKeys;
+
+    function getKeyAfter(key: Key) {
+      let node = keyMap.get(key);
+      if (!node) {
+        return null;
+      }
+  
+      if ((expandedKeys.has(node.key) || node.type !== 'item') && node.firstChildKey != null) {
+        node = keyMap.get(node.firstChildKey);
+        while (node && node.type === 'content' && node.nextKey != null) {
+          node = keyMap.get(node.nextKey);
+        }
+        return node ? node.key : null;
+      }
+  
+      while (node) {
+        if (node.nextKey != null) {
+          return node.nextKey;
+        }
+  
+        if (node.parentKey != null) {
+          node = keyMap.get(node.parentKey);
+        } else {
+          return null;
+        }
+      }
+  
+      return null;
+    }
+
+    let firstKey = this.getFirstKey();
+    let node = firstKey ? keyMap.get(firstKey) : null;
+    for (let i = 0; i < idx; i++) {
+      if (node) {
+        let keyAfter = getKeyAfter(node.key);
+        node = keyAfter ? keyMap.get(keyAfter) : null;
+      }
+    }
+    return node as Node<T>;
   }
 
   getFirstKey() {
-    return this.flattenedRows[0]?.key;
+    let node = this.keyMap.get(this.firstKey);
+    if (!node) {
+      return null;
+    }
+
+    // Skip over any nodes that aren't an item node (e.g. section or header node)
```

Comment:
I wonder if this should be the job of the collection or the keyboard delegate? In BaseCollection we don't skip any nodes here. I kinda think collections should traverse everything and everything downstream (keyboard delegate, dnd, etc.) should skip over things it doesn't care about.